synchronize nested ol.layer.group Items

jbo023 opened the pull request:
Hi, 
we are at the moment working on a project to integrate 3D in the 2D gis portal of Hamburg. For this we are using ol-cesium, but we had to add some functionality and now want to integrate the changes in the project. 
For better review purposes i will create several pull requests. 

Here we changed the code to also synchronize nested ol.layer.group Items. 


Link to the hamburg application: https://www.geoportal-hamburg.de/Geoportal/geo-online/
Link to the Hamburg application source code: https://bitbucket.org/lgv-g12/lgv

This work is paid by the Agency for Geoinformation and Surveying (LGV) Hamburg
http://www.hamburg.de/bsw/whoweare/


Jannes

gberaudo commented:
Hi @jbo023 thanks for upstreaming your changes, I am looking forward to it.
I will not have time to look at your PR(s) before the middle of next week, but as soon I get time I will definitely come back to you.

gberaudo commented:
Hi @jbo023, have you already signed the OpenLayers contributor license agreement? If not it is available at https://github.com/openlayers/openlayers/blob/master/CONTRIBUTING.md#contributor-license-agreement.

jbo023 commented:
I don't think we already had a cla, so i send a signed corporation form to the specified email.

gberaudo commented:
Hi @jbo023, I noticed your compagny signed the OpenLayers CLA. Great!

Regarding this PR, could you please give me some context of what you are trying to achieve? There is already support for ol.layer.Group in OL-Cesium, is it missing something?

jbo023 commented:
@gberaudo before these changes ol-cesium only added the layers of a ol.layer.group but did not synchronize the visibility/opacity state for the complete hierarchy. You can easily check in the commited example if you revert commit https://github.com/openlayers/ol-cesium/pull/494/commits/1e16f56ce0672aea49e64b51ffe4a2706599b85f

olLayerList is now an array of the layer and all of its ol.layer.group parents

gberaudo commented on src/olcs/core.js:
```diff
@@ -407,19 +407,25 @@ olcs.core.tileLayerToImageryLayer = function(olLayer, viewProj) {
- * @param {!ol.layer.Base} olLayer
- * @param {!Cesium.ImageryLayer} csLayer
+ * @param {Array.<!ol.layer.Base>} olLayerList
+ * @param {!Cesium.ImageryLayer | !Cesium.Primitive} csLayer
```
Do you need to change the signature here?

gberaudo commented on src/olcs/abstractsynchronizer.js:
```diff
@@ -98,27 +98,32 @@ olcs.AbstractSynchronizer.prototype.orderLayers = function() {
 olcs.AbstractSynchronizer.prototype.addLayers_ = function(root) {
-  /** @type {Array.<!ol.layer.Base>} */
-  const fifo = [root];
+  /** @type {Array.<Array.<!ol.layer.Base>>} */
```
Using an array of array is disturbing. Could you introduce a new `olcsx.LayerWithParents` type (in olcsx.js) and use it here?

It would look like:
```js
/**
 * @typedef {{
 *   layer: {ol.layer.Layer},
 *   parents: {Array<ol.layer.Group>}
 * }}
 */
olcsx.LayerWithParents;
```

fredj commented on src/olcs/rastersynchronizer.js:
```diff
@@ -94,23 +94,25 @@ olcs.RasterSynchronizer.prototype.convertLayerToCesiumImageries = function(olLay
-          }
-        }));
+    olLayerList.forEach((olLayerItem) => {
+      listenKeyArray.push(olLayerItem.on(['change:opacity', 'change:visible'], (e) => {
```
very very minor: the `e` option is not used, it can be removed

jbo023 commented:
thanks for the review, we had a bit of intergeo stress last week, so i only found the time today. 
I added an entry to the changelog, and modified the code according to your suggestions.

gberaudo commented on CHANGES.md:
```diff
@@ -3,9 +3,15 @@
 # v 1.32
 
+* Breaking changes
+  * Changed dAbstractSynchronizer.createSingleLayerCounterparts Plugin API
```
Typo: `dAbstractSynchronizer`.

gberaudo commented on CHANGES.md:
```diff
@@ -3,9 +3,15 @@
 
+* Breaking changes
+  * Changed dAbstractSynchronizer.createSingleLayerCounterparts Plugin API
+    in existing plugins the function has to be modified.
```
It now takes an object `{layer: !ol.layer.Base, parents: Array<ol.layer.Group>}`. The parents are ordered with the first parent coming first and the last parent coming last.

jbo023 commented:
@gberaudo 

I rebased to the current master, do you need anything else for this to merge ?

gberaudo commented:
@jbo023, there is now a new conflict :(
I tried resolving it through the github interface but the PR is still marked as having conflicts.
Could you try to rebase once more?

jbo023 commented:
@gberaudo yeah i will do another rebase :)

jbo023 commented:
@gberaudo done

fcornelius commented:
Awesome! Any chance we can also get #498 to merge now?
